File: solutions/scala/059-decode-ways.scala
/**
 * Problem 59: Decode Ways (LeetCode 91)
 * Difficulty: Med
 * Language: Scala
 */
def numDecodings(s: String): Int = {
  val n = s.length
  (0 until n).reverse.foldLeft((1, 0)) { case ((dp1, dp2), i) =>
    if (s(i) == '0') (0, dp1)
    else {
      val one = dp1
      val two = if (i + 1 < n && s.substring(i, i + 2).toInt <= 26) dp2 else 0
      (one + two, dp1)
    }
  }._1
}

File: solutions/clojure/059-decode-ways.clj
;; Problem 59: Decode Ways (LeetCode 91)
;; Difficulty: Med
;; Language: Clojure
;; 
(defn num-decodings [s]
  (let [n (count s)]
    (first
      (reduce (fn [[dp1 dp2] i]
                (if (= (nth s i) \0) [0 dp1]
                  (let [one dp1
                        two (if (and (< (inc i) n)
                                     (<= (Integer/parseInt (subs s i (+ i 2))) 26))
                               dp2 0)]
                    [(+ one two) dp1])))
              [1 0]
              (range (dec n) -1 -1)))))
